gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProc.java

```diff
@@ -0,0 +1,199 @@
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration.proc;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+
+import com.codahale.metrics.Timer;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.metrics.ServiceMetricNames;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.runtime.api.DagActionStore;
+import org.apache.gobblin.service.ExecutionStatus;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.flowgraph.DagNodeId;
+import org.apache.gobblin.service.modules.orchestration.DagManagementStateStore;
+import org.apache.gobblin.service.modules.orchestration.DagManagerUtils;
+import org.apache.gobblin.service.modules.orchestration.task.ReevaluateDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+import org.apache.gobblin.service.monitoring.FlowStatusGenerator;
+import org.apache.gobblin.service.monitoring.JobStatus;
+import org.apache.gobblin.service.monitoring.JobStatusRetriever;
+
+
+/**
+ * An implementation for {@link DagProc} that launches a new job if there exists a job whose pre-requisite jobs are
+ * completed successfully. If there are no more jobs to run and no job is running for the Dag, it cleans up the Dag.
+ */
+@Slf4j
+public class ReevaluateDagProc extends DagProc<Optional<Dag.DagNode<JobExecutionPlan>>, Void> {
+  private final JobStatusRetriever jobStatusRetriever;
+  private final Timer jobStatusPolledTimer;
+  private final DagNodeId dagNodeId;
+  private JobStatus jobStatus;
+
+  public ReevaluateDagProc(ReevaluateDagTask reEvaluateDagTask, JobStatusRetriever jobStatusRetriever) {
+    super(reEvaluateDagTask);
+    this.jobStatusRetriever = jobStatusRetriever;
+    this.jobStatusPolledTimer = metricContext.timer(ServiceMetricNames.JOB_STATUS_POLLED_TIMER);
+    this.dagNodeId = getDagNodeId();
+  }
+
+  @Override
+  protected Optional<Dag.DagNode<JobExecutionPlan>> initialize(DagManagementStateStore dagManagementStateStore)
+      throws IOException {
+    Optional<Dag.DagNode<JobExecutionPlan>> dagNode = dagManagementStateStore.getDagNode(this.dagNodeId);
+    if (!dagNode.isPresent()) {
+      log.error("DagNode not found for a Reevaluate DagAction with dag node id {}", this.dagNodeId);
+      return Optional.empty();
+    }
+    this.jobStatus = DagManagerUtils.pollJobStatus(dagNode.get(), this.jobStatusRetriever, this.jobStatusPolledTimer).get();
+    ExecutionStatus executionStatus = ExecutionStatus.valueOf(jobStatus.getEventName());
+    if (!FlowStatusGenerator.FINISHED_STATUSES.contains(executionStatus.name())) {
+      log.warn("Job status for dagNode {} is {}. Expected Statuses are {}", dagNodeId, executionStatus, FlowStatusGenerator.FINISHED_STATUSES);
+      return Optional.empty();
+    }
+    setStatus(dagManagementStateStore, dagNode.get(), executionStatus);
```

Comment:
job status retriever sets the status in jobStatusStore.
here, we are updating the status in Dag which is stored in DMSS.
we can consider merging them both in future.

Reply:
JobStatusStore is the SoT. this method is copying the status from JobStatusStore to DMSS. and the status is queried through DMSS only by this class. What can lead to errors in near term?
Merging dag stores, various maps and (now) job status store will not be straightforward; I will have to come up with a schema(s) that can provide atomic updates to any information about the dag.

Reply:
Oh and we could have just avoided this copy, if `DagManagerUtils.getExecutionStatus` would be looking into the SoT (JobStatusStore) instead of looking into the dag, to find the status.